File: AgoraChat-ObjectiveC/AgoraChat/Code/Controllers/Groups/ACDTextViewController.m
//
//  EMTextViewController.m
//  ChatDemo-UI3.0
//
//  Created by XieYajie on 2019/1/17.
//  Copyright © 2019 XieYajie. All rights reserved.
//

#import "ACDTextViewController.h"

#import "ACDTextView.h"

@interface ACDTextViewController ()<UITextViewDelegate>

@property (nonatomic, strong) NSString *originalString;
@property (nonatomic, strong) NSString *placeholder;
@property (nonatomic) BOOL isEditable;

@property (nonatomic, strong) ACDTextView *textView;

@end

@implementation ACDTextViewController

- (instancetype)initWithString:(NSString *)aString
                   placeholder:(NSString *)aPlaceholder
                    isEditable:(BOOL)aIsEditable
{
    self = [super init];
    if (self) {
        _originalString = aString;
        _placeholder = aPlaceholder;
        _isEditable = aIsEditable;
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = self.navTitle;
    
    [self _setupSubviews];
}

#pragma mark - Subviews

- (void)_setupSubviews
{

    self.navigationItem.leftBarButtonItem = [ACDUtil customBarButtonItem:@"Cancel" action:@selector(goBack) actionTarget:self];
    
    if (self.isEditable) {
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"common.done", nil) style:UIBarButtonItemStylePlain target:self action:@selector(doneAction)];
        self.navigationItem.rightBarButtonItem.tintColor = ButtonEnableBlueColor;
    }
    
    self.view.backgroundColor = UIColor.whiteColor;
    UIView *bgView = [[UIView alloc] init];
    bgView.backgroundColor = [UIColor whiteColor];
    [self.view addSubview:bgView];
    [bgView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
    
    self.textView = [[ACDTextView alloc] init];
    self.textView.delegate = self;
    self.textView.backgroundColor = [UIColor clearColor];
    self.textView.font = [UIFont systemFontOfSize:16];
    if (!self.isEditable)
        self.textView.placeholder = NSLocalizedString(@"editRight", nil);
    else
        self.textView.placeholder = self.placeholder;
    self.textView.returnKeyType = UIReturnKeyDone;
    if (self.originalString && ![self.originalString isEqualToString:@""]) {
        self.textView.text = self.originalString;
    }
    self.textView.editable = self.isEditable;
    [self.view addSubview:self.textView];
    [self.textView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.equalTo(bgView);
        make.top.equalTo(bgView).offset(5);
        make.left.equalTo(bgView).offset(10);
    }];
}

- (void)goBack {
    [self dismissViewControllerAnimated:YES completion:nil];
    
}

#pragma mark - UITextViewDelegate

- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
{
    if ([text isEqualToString:@"\n"]) {
        [textView resignFirstResponder];
        return NO;
    }
    
    return YES;
}

#pragma mark - Action

- (void)doneAction
{
    [self.view endEditing:YES];
    
    BOOL isPop = YES;
    if (_doneCompletion) {
        isPop = _doneCompletion(self.textView.text);
    }
    
    if (isPop) {
        [self goBack];
    }
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Groups/ACDTextViewController.swift
//
//  ACDTextViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/27.
//

import UIKit

class ACDTextViewController: UIViewController {

    @IBOutlet weak var textView: UITextView!
    
    private let text: String
    private let placeholder: String
    private let isEditable: Bool
    
    var doneCompletion: ((_ text: String) -> Bool)?
    
    init(text: String, placeholder: String, isEditable: Bool) {
        self.text = text
        self.placeholder = placeholder
        self.isEditable = isEditable
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let cancelBtn = UIButton(type: .custom)
        cancelBtn.frame = CGRect(x: 0, y: 0, width: 50, height: 40)
        cancelBtn.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        cancelBtn.setTitleColor(Color_154DFE, for: .normal)
        cancelBtn.setTitle(LocalizedString.Cancel, for: .normal)
        cancelBtn.addTarget(self, action: #selector(goBack), for: .touchUpInside)
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(customView: cancelBtn)
        
        if self.isEditable {
            self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Done".localized, style: .plain, target: self, action: #selector(doneAction))
        }
        self.textView.text = self.text
        self.textView.becomeFirstResponder()
    }
    
    @objc private func goBack() {
        self.navigationController?.popViewController(animated: true)
    }
    
    @objc private func doneAction() {
        self.view.endEditing(true)
        if self.doneCompletion?(self.textView.text) ?? true {
            self.goBack()
        }
    }
}

extension ACDTextViewController: UITextViewDelegate {
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        if text == "\n" {
            textView.resignFirstResponder()
            return false
        }
        return true
    }
}
